src/aspire/source/micrograph.py

```diff
@@ -0,0 +1,306 @@
+                or boundary > self.micrograph_size // 2
+            ):
+                raise RuntimeError("Illegal boundary value.")
+            self.boundary = boundary
+
+        if interparticle_distance is None:
+            self.interparticle_distance = self.particle_box_size
+        else:
+            self.interparticle_distance = interparticle_distance
+
+        # Create the radial mask for each center
+        g2d = grid_2d(int(2 * self.interparticle_distance), normalized=False)
+        radial_mask = g2d["r"] <= self.interparticle_distance
+        self.grid_x = g2d["x"].astype(int)[radial_mask]
+        self.grid_y = g2d["y"].astype(int)[radial_mask]
+
+        # Calculate the proper padding and create the mask
+        self.pad = int(max(self.particle_box_size, self.interparticle_distance))
+        self._mask = np.full(
+            (
+                int(self.micrograph_size + 2 * self.pad),
+                int(self.micrograph_size + 2 * self.pad),
+            ),
+            False,
+            dtype=bool,
+        )
+        self._mask_boundary = self.pad + self.boundary
+        self._mask[
+            self._mask_boundary : -1 * self._mask_boundary,
+            self._mask_boundary : -1 * self._mask_boundary,
+        ] = True
+
+        # Create the centers
+        self.centers = np.zeros(
+            (self.micrograph_count, self.particles_per_micrograph, 2), dtype=int
+        )
+        self._fail_limit = pass_threshold * self.micrograph_count
+        self._fail_count = 0
+        for i in range(self.micrograph_count):
+            self.centers[i] = self._create_centers(i)
+
+        self._clean_micrographs_accessor = _ImageAccessor(
+            self._clean_micrographs, self.micrograph_count
+        )
+        self._micrographs_accessor = _ImageAccessor(
+            self._micrographs, self.micrograph_count
+        )
+
+        self.images = _ImageAccessor(self._images, self.total_particle_count)
+
+    def _create_centers(self, micrograph):
+        """
+        Creates centers for the given micrograph if the fail threshold isn't met.
+
+        param micrograph: The ID of the micrograph.
+        return: A numpy array containing the generated centers.
+        """
+        while self._fail_count < self._fail_limit:
+            try:
+                self._reset_mask()
+                centers = np.zeros((self.particles_per_micrograph, 2))
+                for i in range(self.particles_per_micrograph):
+                    x, y = self._generate_center(micrograph)
+                    centers[i] = np.array([x, y])
+                return centers
+            except RuntimeError:
+                self._fail_count += 1
+        else:
+            raise RuntimeError(
+                "Micrograph generation failures exceeded limit. This can happen if constraints are too strict. Consider adjusting pass_threshold, micrograph_size, particle_count, or interparticle_distance."
```

Comment:
Out of these parameters, I think it makes more sense for users to adjust `micrograph_size`, `particle_count`, and `interparticle_distance`. I would argue that `pass_threshold` should perhaps be an internal constant (set to a very low value), then the users will adjust the other parameters accordingly.

Another potential direction would be to relax `particle_count` and replace it with something like `max_particle_count`. That is, we try to place that number of particles, but if they don't fit, we accept a lower number (and maybe issue a warning with the same message as what we have here).

Reply:
I'll make a hard-coded value right now, and make an issue for making particle count non-uniform in the future.